interop-reactiveStreams/shared/src/main/scala/scalaz/zio/interop/reactiveStreams/StreamPublisher.scala

```diff
@@ -0,0 +1,55 @@
+package scalaz.zio.interop.reactiveStreams
+
+import org.reactivestreams.{ Publisher, Subscriber, Subscription }
+import scalaz.zio.stream.{ Sink, Stream }
+import scalaz.zio.{ Queue, Runtime, UIO, ZIO }
+
+class StreamPublisher[R, E <: Throwable, A](
+  stream: Stream[R, E, A],
+  runtime: Runtime[R]
+) extends Publisher[A] {
+
+  override def subscribe(subscriber: Subscriber[_ >: A]): Unit =
+    if (subscriber == null) {
+      throw new NullPointerException("Subscriber must not be null.")
+    } else {
+      runtime.unsafeRunAsync_(
+        for {
+          demand  <- Queue.unbounded[Long]
+          _       <- UIO(subscriber.onSubscribe(createSubscription(subscriber, demand)))
+          control = Stream.fromQueue(demand).flatMap(n => Stream.unfold(n)(n => if (n > 0) Some(((), n - 1)) else None))
+          fiber <- stream
+                    .zipWith(control) {
+                      case (Some(a), Some(_)) =>
```

Comment:
Ah this is nice. You still need to suspend the side-effects though with a `Task`; e.g. 
```scala
stream.zipWith(control) {
  case (Some(a), Some(_)) =>
    Some(Task(subscriber.onNext(a)))
  // rest of the cases
}.mapM(identity)
```

This is because unsuspended side effects like these may behave unexpectedly if `zipWith` is refactored. At this point it might be more succinct to just do `stream.foreach { a => demand.take *> Task(subscriber.onNext(a)) }`.

Reply:
Suspension done for the first case. No clue if it is possible to achieve for the second case. I tend to just live with the current state for now.

`demand.take` alone does not do the job. I still need to unfold the demand. That would leave me with another fromQueue ... toQueue construct (maybe Sink.queue). I'm up for suggestions. Maybe something rather complex using mapAccumM?

Reply:
Ah yeah, that's true. How about just pushing that complexity to the other side? In `createSubscription` you could offer `n` tokens to the queue.

Reply:
n could be Long.MaxValue. I don't have enough RAM for that approach. ;-)